File: variant-a/java-service/src/main/java/com/flashsale/api/controller/HealthController.java
package com.flashsale.api.controller;

import com.flashsale.api.service.CampaignMemoryAllocator;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.UUID;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Health check endpoint for load balancer.
 */
@RestController
public class HealthController {

    private final CampaignMemoryAllocator allocator;
    
    // Benchmark constants
    private static final UUID BENCHMARK_CAMPAIGN_ID = UUID.fromString("00000000-0000-0000-0000-000000000001");
    private static final UUID BENCHMARK_SPU_ID = UUID.fromString("00000000-0000-0000-0000-000000000003");
    private static final UUID BENCHMARK_SKU_ID = UUID.fromString("00000000-0000-0000-0000-000000000002");
    
    private volatile boolean benchmarkLoaded = false;
    private final ReentrantLock benchmarkLock = new ReentrantLock();

    public HealthController(CampaignMemoryAllocator allocator) {
        this.allocator = allocator;
    }

    /**
     * Health check endpoint.
     * Returns plain text "200 OK" following BoA internal pattern.
     * Supports both GET and HEAD methods.
     */
    @RequestMapping(value = "/health", method = {RequestMethod.GET, RequestMethod.HEAD})
    public ResponseEntity<String> health() {
        return ResponseEntity.ok("200 OK");
    }

    /**
     * Benchmark endpoint to emulate Order creation (Fast Path only).
     * Emulates infinite inventory (10M items) to test raw allocation overhead without Redis I/O.
     */
    @RequestMapping(value = "/health2", method = RequestMethod.GET)
    public ResponseEntity<String> healthBenchmark() {
        // Lazy load benchmark campaign
        if (!benchmarkLoaded) {
            benchmarkLock.lock();
            try {
                if (!benchmarkLoaded) {
                    allocator.loadCampaign(
                        BENCHMARK_CAMPAIGN_ID,
                        BENCHMARK_SPU_ID,
                        Collections.singletonMap(BENCHMARK_SKU_ID, 10_000_000), // 10 Million items
                        new BigDecimal("100.00"),
                        new BigDecimal("200.00"),
                        20.0
                    );
                    benchmarkLoaded = true;
                }
            } finally {
                benchmarkLock.unlock();
            }
        }

        // Perform exactly one reservation (Fast Path simulation)
        CampaignMemoryAllocator.ReservationResult result = allocator.reserveItem(
            BENCHMARK_CAMPAIGN_ID,
            BENCHMARK_SKU_ID
        );

        if (result.success) {
            return ResponseEntity.ok("200 OK");
        } else {
            return ResponseEntity.internalServerError().body("Benchmark allocation failed: " + result.priceType);
        }
    }
}

File: variant-a/csharp-service/Controllers/HealthController.cs
using Microsoft.AspNetCore.Mvc;
using FlashSaleAPI.Services;
using System;
using System.Collections.Generic;

namespace FlashSale.Api.Controllers
{
    /// <summary>
    /// Health check endpoint for load balancer.
    /// </summary>
    [ApiController]
    public class HealthController : ControllerBase
    {
        private readonly CampaignMemoryAllocator _allocator;
        private static readonly Guid BenchmarkCampaignId = Guid.Parse("00000000-0000-0000-0000-000000000001");
        private static readonly Guid BenchmarkSkuId = Guid.Parse("00000000-0000-0000-0000-000000000002");
        private static bool _benchmarkLoaded = false;
        private static readonly object _benchmarkLock = new object();

        public HealthController(CampaignMemoryAllocator allocator)
        {
            _allocator = allocator;
        }

        /// <summary>
        /// Health check endpoint.
        /// Returns plain text "200 OK" following BoA internal pattern.
        /// Supports both GET and HEAD methods.
        /// </summary>
        [HttpGet("/health")]
        [HttpHead("/health")]
        public IActionResult Health()
        {
            return Ok("200 OK");
        }

        /// <summary>
        /// Benchmark endpoint to emulate Order creation (Fast Path only).
        /// Emulates infinite inventory (10M items) to test raw allocation overhead without Redis I/O.
        /// </summary>
        [HttpGet("/health2")]
        public async Task<IActionResult> Health2()
        {
            // Lazy load benchmark campaign with massive inventory
            if (!_benchmarkLoaded)
            {
                lock (_benchmarkLock)
                {
                    if (!_benchmarkLoaded)
                    {
                        var skuAllocations = new Dictionary<Guid, int>
                        {
                            { BenchmarkSkuId, 10_000_000 } // 10 Million items
                        };

                        _allocator.LoadCampaign(
                            BenchmarkCampaignId,
                            Guid.NewGuid(),
                            skuAllocations,
                            100m,
                            200m,
                            20 // 20% watermark
                        );
                        _benchmarkLoaded = true;
                    }
                }
            }

            // Perform exactly one reservation (Fast Path simulation)
            var result = await _allocator.ReserveItemAsync(BenchmarkCampaignId, BenchmarkSkuId);

            if (result.Success)
            {
                return Ok("200 OK");
            }
            else
            {
                return StatusCode(500, $"Benchmark allocation failed: {result.PriceType}");
            }
        }

        /// <summary>
        /// Benchmark: POST + JSON parsing + allocation + JSON response
        /// Isolates JSON overhead from full order path.
        /// </summary>
        [HttpPost("/health3")]
        public async Task<IActionResult> Health3([FromBody] Health3Request request)
        {
            // Lazy load benchmark campaign
            if (!_benchmarkLoaded)
            {
                lock (_benchmarkLock)
                {
                    if (!_benchmarkLoaded)
                    {
                        var skuAllocations = new Dictionary<Guid, int>
                        {
                            { BenchmarkSkuId, 10_000_000 }
                        };
                        _allocator.LoadCampaign(BenchmarkCampaignId, Guid.NewGuid(), skuAllocations, 100m, 200m, 20);
                        _benchmarkLoaded = true;
                    }
                }
            }

            // Allocation
            var result = await _allocator.ReserveItemAsync(BenchmarkCampaignId, BenchmarkSkuId);

            if (result.Success)
            {
                return Ok(new Health3Response { Success = true, Price = result.Price });
            }
            return StatusCode(500, new { error = result.PriceType });
        }

        /// <summary>
        /// Benchmark: Full order simulation using REAL campaign (with refills)
        /// </summary>
        [HttpPost("/health4")]
        public async Task<IActionResult> Health4([FromBody] Health3Request request)
        {
            // Use real campaign
            var campaignId = Guid.Parse("99999999-8888-7777-6666-555544443333");
            var skuId = Guid.Parse("11111111-2222-3333-4444-555555555555");

            var result = await _allocator.ReserveItemAsync(campaignId, skuId);

            if (result.Success)
            {
                // Build payload like real order (but don't queue)
                var payload = new Dictionary<string, object>
                {
                    ["order_number"] = $"ORD-{DateTime.UtcNow.Ticks}",
                    ["sku_id"] = skuId.ToString(),
                    ["price"] = result.Price
                };
                return Ok(new Health3Response { Success = true, Price = result.Price });
            }
            return StatusCode(500, new { error = result.PriceType });
        }
    }

    public class Health3Request
    {
        public string? Email { get; set; }
        public int Quantity { get; set; } = 1;
    }

    public class Health3Response
    {
        public bool Success { get; set; }
        public decimal Price { get; set; }
    }
}
